soweego/wikidata/api_requests.py

```diff
@@ -36,9 +34,272 @@
+
+    for bucket in qid_buckets:
+        entities = _sanity_check(bucket, request_params)
+
+        if entities is None:
+            continue
+
+        for qid in entities:
+            entity = entities[qid]
+
+            # Sitelinks
+            yield _yield_sitelinks(entity, qid, no_sitelinks_count)
+
+            claims = entity.get('claims')
+            if claims:
+                # Third-party links
+                yield _yield_expected_values(
+                    qid, claims, url_pids, no_links_count
+                )
+
+                # External ID links
+                yield _yield_ext_id_links(
+                    ext_id_pids_to_urls, claims, qid, no_ext_ids_count
+                )
+            else:
+                LOGGER.info('No claims for QID %s', qid)
+
+    LOGGER.info(
+        'QIDs: got %d with no sitelinks, '
+        '%d with no third-party links, '
+        '%d with no external ID links',
+        no_sitelinks_count,
+        no_links_count,
+        no_ext_ids_count,
+    )
+
+
+def get_data_for_linker(
+    catalog: str,
+    entity: str,
+    qids: Set[str],
+    url_pids: Set[str],
+    ext_id_pids_to_urls: Dict,
+    qids_and_tids: Dict,
+    fileout: TextIO,
+) -> None:
+    """Collect relevant data for linking Wikidata to a given catalog.
+    Dump the result to a given output stream.
+
+    This function uses multithreaded parallel processing.
+
+    :param catalog: ``{'discogs', 'imdb', 'musicbrainz'}``.
+      A supported catalog
+    :param entity: ``{'actor', 'band', 'director', 'musician', 'producer',
+      'writer', 'audiovisual_work', 'musical_work'}``.
+      A supported entity
+    :param qids: a set of QIDs
+    :param url_pids: a set of PIDs holding URL values.
+      Returned by :py:func:`soweego.wikidata.sparql_queries.url_pids`
+    :param ext_id_pids_to_urls: a
+      ``{PID: {formatter_URL: formatter_regex} }`` dict.
+      Returned by
+      :py:func:`soweego.wikidata.sparql_queries.external_id_pids_and_urls`
+    :param fileout: a file stream open for writing
+    :param qids_and_tids: a ``{QID: {'tid': {catalog_ID_set} }`` dict.
+      Populated by
+      :py:func:`soweego.commons.data_gathering.gather_target_ids`
+    """
+    qid_buckets, request_params = _prepare_request(
+        qids, 'labels|aliases|descriptions|sitelinks|claims'
+    )
+
+    # Catalog-specific data needs
+    if catalog in constants.REQUIRE_OCCUPATION.keys():
+        needs_occupation = entity in constants.REQUIRE_OCCUPATION[catalog]
+    else:
+        needs_occupation = False
+    needs_genre = entity in constants.REQUIRE_GENRE
+    needs_publication_date = entity in constants.REQUIRE_PUBLICATION_DATE
+
+    # Initialize 7 counters to 0
+    # Indices legend:
+    # 0 = claims
+    # 1 = labels
+    # 2 = aliases
+    # 3 = descriptions
+    # 4 = sitelinks
+    # 5 = third-party URLs
+    # 6 = third-party IDs
+    counters = [0] * 7
```

Comment:
Maybe this can also be accomplished in a cleaner way using a `namedtuple` ?

Reply:
I see what you mean, but we need to update the counters and the `namedtuple` doesn't allow so.